engine/src/main/java/org/camunda/bpm/engine/impl/cfg/auth/DefaultAuthorizationProvider.java

```diff
@@ -292,10 +294,53 @@
           updateAuthorization(historyAuthorization, userId, groupId, HISTORIC_TASK,
               taskId, HistoricTaskPermissions.READ);
 
+      String rootProcessInstanceId = getRootProcessInstanceId(task);
+
+      historyAuthorization.setRootProcessInstanceId(rootProcessInstanceId);
+
+      if (isHistoryRemovalTimeStrategyStart()) {
+        HistoryEvent rootProcessInstance = findHistoricProcessInstance(rootProcessInstanceId);
+
+        Date removalTime = null;
+        if (rootProcessInstance != null) {
+          removalTime = rootProcessInstance.getRemovalTime();
+
+        }
+
+        historyAuthorization.setRemovalTime(removalTime);
+      }
+
```

Comment:
The logic in the lines above (setting root instance id and removal time) is only necessary when the authorization is created, not when it is updated. I think it would be good to only perform it in the creation case to avoid unnecessary queries for the root process instance.

Reply:
When we only do it on creation, we miss the following case:
1. A user creates a historic task authorization manually
2. The user (1) realizes later that there is a typo in the task id or (2) wants to change the authorization from ANY to a concrete task id
3. No root process instance id is assigned

This is counter-intuitive for the user.

Reply:
Isn't that case covered by the changes you made in `SaveAuthorizationCmd` via CAM-11617? When the user fixes the authorization, then it automatically receives the root instance id.

Reply:
Sorry, I did not pay enough attention to the exact code and mixed both places up. I think you are right.